e2e-test-server/src/main/java/com/google/cloud/opentelemetry/endtoend/PubSubPushServer.java

```diff
@@ -0,0 +1,221 @@
+ *   <li>Validate incoming HTTP requests and parse them into {@link PubsubMessage}s.
+ *   <li>Hand off the parsed {@link PubsubMessage}s to a specified {@link PubSubMessageHandler}.
+ * </ul>
+ */
+public class PubSubPushServer implements PubSubServer {
+
+  private static final String POST_REQUEST = "POST";
+
+  private final int port;
+  private final HttpServer httpServer;
+  private final HttpHandler rootRequestHandler;
+  private final PubSubMessageHandler pubsubMessageHandler;
+
+  /**
+   * Public constructor for the {@link PubSubPushServer}.
+   *
+   * @param port The port on which the HTTP server should listen for incoming requests.
+   * @param pubSubMessageHandler The {@link PubSubMessageHandler} responsible for handling the
+   *     incoming HTTP requests.
+   */
+  public PubSubPushServer(int port, PubSubMessageHandler pubSubMessageHandler) {
+    this.port = port;
+    this.pubsubMessageHandler = pubSubMessageHandler;
+    this.rootRequestHandler = createRootRequestHandler();
+    this.httpServer = createHttpServer();
+  }
+
+  @Override
+  public void start() {
+    this.httpServer.start();
+  }
+
+  @Override
+  public void close() {
+    httpServer.stop(60);
+    pubsubMessageHandler.cleanupMessageHandler();
+  }
+
+  private PubsubMessage parseIncomingMessage(HttpExchange httpExchange) {
+    InputStreamReader inputStreamReader =
+        new InputStreamReader(httpExchange.getRequestBody(), StandardCharsets.UTF_8);
+    JsonElement jsonRoot = JsonParser.parseReader(inputStreamReader);
+    String msgStr = jsonRoot.getAsJsonObject().get("message").toString();
+    Gson gson = new Gson();
+    Message message = gson.fromJson(msgStr, Message.class);
+    return PubsubMessage.newBuilder().putAllAttributes(message.getAttributes()).build();
+  }
+
+  private HttpServer createHttpServer() {
+    HttpServer httpServer = null;
+    try {
+      httpServer = HttpServer.create(new InetSocketAddress(this.port), 0);
+      httpServer.createContext("/", this.rootRequestHandler);
+      httpServer.setExecutor(MoreExecutors.directExecutor());
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+    return httpServer;
+  }
+
+  private HttpHandler createRootRequestHandler() {
+    return httpExchange -> {
+      if (httpExchange.getRequestMethod().equals(POST_REQUEST)) {
+        if (!isRequestJSON(httpExchange)) {
+          String response = "Expecting request in JSON format";
+          httpExchange.sendResponseHeaders(400, response.length());
+          OutputStream os = httpExchange.getResponseBody();
+          os.write(response.getBytes(StandardCharsets.UTF_8));
+          os.close();
+          httpExchange.close();
+          return;
+        }
+        PubsubMessage message = parseIncomingMessage(httpExchange);
+        PubSubMessageResponse ack_or_nack = pubsubMessageHandler.handlePubSubMessage(message);
+        String finalResponse = "";
```

Comment:
Is the response body needed or can we just send back the response code?

Reply:
Probably not, I checked the docs again https://docs.oracle.com/javase/8/docs/jre/api/net/httpserver/spec/com/sun/net/httpserver/HttpExchange.html#sendResponseHeaders-int-long- and I think I could use something like - `httpExchange.sendResponseHeaders(200, 0);`